concepts.tex

```diff
@@ -50,24 +57,74 @@
 when violated, make the function non-regular. \enternote A function that returns
 \tcode{void} is necessarily a regular function. \exitnote}
 
+{\color{newclr}
 \pnum
-\newtxt{Expressions appearing in a \techterm{requires-expression} in this document
+Expressions appearing in a \techterm{requires-expression} in this document
 are required to be equality preserving, except for those annotated with the comment
 ``not equality preserving.'' An expression so annotated may be equality preserving,
-but it is not required to be so.}
+but is not required to be so.
+
+\pnum
+An expression that may alter the value of one or more of its inputs in a manner
+observable to equality preserving expressions is said to modify those inputs.
+This document uses a notational convention to specify which expressions modify what
+inputs: except where otherwise specified, an expression operand that is a
+non-constant lvalue or rvalue may be modified. Similarly, operands that are
+constant lvalues or rvalues must not be modified.
+
+\pnum
+Where a concept declares an expression that is non-modifying for some constant
+lvalue operand, additional variants of that expression that accept a non-constant
+lvalue or (possibly constant) rvalue for the given operand are also required with
+identical semantics except where such an expression variant is explicitly required
+with differing semantics. The extent to which the library implementation validates
+these implicit expression variants is unspecified.
+
+\enterexample
+\begin{codeblock}
+template <class T>
+concept bool C() {
+  return requires (T a, T b, const T c, const T d) {
+    c == d;       // \#1
+    a = move(b);  // \#2
+    a = c;        // \#3
+  };
+}
+\end{codeblock}
+Expression \#1 does not modify either of its operands, \#2 modifies both of its
+operands, and \#3 modifies only its first operand \tcode{a}.
+
+Expression \#1 implicitly requires additional expression variants with semantics
+identical to \tcode{c == d} (including non-modification), as if the expressions
+\begin{codeblock}
+a == d;       a == b;             a == move(b);       a == d;
+c == a;       c == move(a);       c == move(d);
+move(a) == d; move(a) == b;       move(a) == move(b); move(a) == move(d);
+move(c) == b; move(c) == move(b); move(c) == d;       move(c) == move(d);
+\end{codeblock}
+had been declared as well.
+
+Expression \#3 implicitly requires additional expression variants with semantics
+identical to \tcode{a = c} (including non-modification of the second operand),
+as if the expressions \tcode{a = b} and \tcode{a = move(c)} had been declared.
+Expression \#3 does not implicitly require an expression variant with a
+non-constant rvalue second operand, since expression \#2 already specifies exactly
+such an expression explicitly.
+\exitexample
```

Comment:
There's a bit of rationale in https://github.com/CaseyCarter/stl2/issues/6, for what it's worth.